Libraries/MiscDrivers/SRAM/N01S830HA.c

```diff
@@ -0,0 +1,177 @@
+{
+    out[0] = cmd;
+    out[1] = (address >> 16) & 0xFF; // MSB first
+    out[2] = (address >> 8) & 0xFF;
+    out[3] = (address & 0xFF);
+}
+
+inline int _transmit_spi_header(uint8_t cmd, uint32_t address)
+{
+    // SPI reads and writes will always start with 4 bytes.
+    // A command byte, then a 24-bit address (MSB first)
+    uint8_t header[4];
+    _parse_spi_header(cmd, address, header);
+
+    // Transmit header, but keep Chip Select asserted.
+    ERR_CHECK(spi_transmit(header, 4, NULL, 0, false));
+    return E_NO_ERROR;
+}
+
+int N01S830HA_init()
+{
+    ERR_CHECK(spi_init());
+    ERR_CHECK(N01S830HA_exit_quadmode()); // Protect against quad-mode lock-up
+
+    // The first thing we need to do is disable the HOLD function, which
+    // is enabled by default.  There is a hardware (hold pin) and software
+    // (hold bit) component to this.
+
+    // Set the hold pin to the HIGH state.
+    ERR_CHECK(MXC_GPIO_Config(&N01S830HA_hold_pin));
+    MXC_GPIO_OutSet(N01S830HA_hold_pin.port, N01S830HA_hold_pin.mask);
+
+    ERR_CHECK(N01S830HA_write_mode_reg(0b1)); // Disable hold function
+
+    // Now, validate that we were able to write to the mode register
+    // This is the closest thing we have to a "read id" or
+    // communication verification for this SRAM chip.
+    uint8_t mode_reg;
+    ERR_CHECK(N01S830HA_read_mode_reg(&mode_reg));
+    if (mode_reg != 0x01) {
+        return E_NO_DEVICE;
+    }
+
+    // Set burst mode (mode reg bits [7:6])
+    // [7:6] = 0b00 -> word mode
+    // [7:6] = 0b10 -> page mode
+    // [7:6] = 0b01 -> burst mode (*)
+    // [7:6] = 0b11 - > reserved
+
+    // Burst mode allows writing across the 8-bit word and
+    // 32-word page boundaries by automatically incrementing and
+    // wrapping the address pointer.
+    mode_reg &= ~(0b11 << 6);
+    mode_reg |= (0b01 << 6);
+    ERR_CHECK(N01S830HA_write_mode_reg(mode_reg));
+
+    // Validate the mode reg is updated successfully.
+    uint8_t validate = 0;
+    ERR_CHECK(N01S830HA_read_mode_reg(&validate));
+    if (validate != mode_reg) {
+        printf("Expected %i, received %i\n", mode_reg, validate);
+        return E_COMM_ERR;
+    }
+
+    // Re-configure SPI pins now that the hold function has been disabled.
+    // This allows us to use QSPI.
+    ERR_CHECK(spi_init());
+
+    return E_NO_ERROR;
+}
+
+int N01S830HA_enter_quadmode()
+{
+    uint8_t tx_data = CMD_ENABLE_QUAD_IO;
+
+    ERR_CHECK(spi_exit_quadmode());
+    ERR_CHECK(spi_transmit(&tx_data, 1, NULL, 0, true));
+    ERR_CHECK(spi_enter_quadmode());
+
+    g_current_mode = QUAD_MODE;
+
+    return E_NO_ERROR;
+}
+
+int N01S830HA_exit_quadmode()
+{
+    uint8_t tx_data = CMD_RESET_IO;
+
+    ERR_CHECK(spi_enter_quadmode());
+    ERR_CHECK(spi_transmit(&tx_data, 1, NULL, 0, true));
+    ERR_CHECK(spi_exit_quadmode());
+
+    g_current_mode = STANDARD_MODE;
+
+    return E_NO_ERROR;
+}
+
+int N01S830HA_read(uint32_t address, uint8_t *out, unsigned int len)
```

Comment:
Add some small comments to the functions in this file. Some of these function took a hot sec to understand and comments would go a long way.

Reply:
Does this help?  https://github.com/Analog-Devices-MSDK/msdk/pull/802/commits/13544d4dbb30101ef617bd1f604d867c41565181  

I think the rest of the functions are pretty readable